### PR TITLE
Added output variable to Docker task

### PR DESCRIPTION
**Pull request by devedse**
Dear all, since chaining docker commands is not possible like it would be in the normal command line with docker rmi $(docker images -aq) for example, I have added an output variable to docker commands.

That way you can add 2 seperate docker commands that provide input to eachother:

1. Output the variable
image

2. Using the variable
image

Also see some StackOverflow posts from me and other people who are running into issues due to this:

- http://stackoverflow.com/questions/42853895/removing-unused-docker-images-on-a-remote-linux-azure-hosted-docker-instance-giv
- http://stackoverflow.com/questions/41957314/command-nesting-in-vsts-tfs

Steps I used to upload the Docker task to TFS to test it:
1. Install tfx
`npm install -g tfx-cli`
2. Login to tfx
`tfx login -u https://********.visualstudio.com/DefaultCollection -t <fillinyourpersonalaccesstokenhere>`
3. Run npm install in the root
`npm i`
4. Build the Docker task
`node make.js build --task Docker`
5. Remove any previous versions of the task if they still exist (not required most of the times)
`tfx build tasks delete --task-id e28912f1-0114-4464-802a-a3a35437fd16`
6. Upload the build task, not the one in /Tasks/Docker but rather the one in /_build/Tasks/Docker
`tfx build tasks upload --task-path ./_build/Tasks/Docker`
7. Add it to your build process and change the 'Container Registry Type' to 'Container Registry' to have the same functionality as the old task
8. Add an output variable + custom condition if required
9. ...
10. Profit :)

P.S.: I first added this changerequest to the vsts-docker repository but got informed I should add it here. Link to old pull request: https://github.com/Microsoft/vsts-docker/pull/44

### CODE REVIEW

**Comment by devedse** [comment]
Anyone got an update on this?

File: Tasks/Docker/containerconnection.ts [hunk 1]
```diff
@@ -37,13 +37,23 @@ export default class ContainerConnection {
...
         var errlines = [];
+        var result = "";
+        command.on("stdout", data => {
+            var dataAsString = data + "";
+            if (result.length > 0) {
+                result += " ";
+            }
+            // Add all stdout lines to the result string, but replace newline by a space because passing the result with newlines to a second Docker command doesn't work.
+            // E.g. if you first do 'docker images -aq', and want to pass that to 'docker rmi', it needs to get it's output split by spaces instead of newlines.
+            result += dataAsString.replace(/[\r\n]+/g, " ");
```
Comment (jikuma) [inline Tasks/Docker/containerconnection.ts]: I think this should not be part of this task. We can have a inline PowerShell or shell script task to manipulate the value of the variable.

File: Tasks/Docker/task.json [hunk 2]
```diff
@@ -274,6 +274,13 @@
             "defaultValue": "$(System.DefaultWorkingDirectory)",
             "helpMarkDown": "Working directory for the Docker command.",
```
Comment (jikuma) [inline Tasks/Docker/task.json]: With every change, we need to bump up the task minor version.

File: Tasks/Docker/containerconnection.ts (hunk 1 above)
Comment (devedse) [inline Tasks/Docker/containerconnection.ts]: Hi jikuma, I don't exactly understand what you mean by this, should I simply remove the .replace here and do the removal of '/r/n' in a seperate PowerShell script?

File: Tasks/Docker/task.json (hunk 2 above)
Comment (devedse) [inline Tasks/Docker/task.json]: I increased the minor version.

**Comment by chrisrpatterson** [comment]
I would like to wait for our formal output variable support to be in place before we integrate this feature.

**Comment by bansalaseem** [comment]
@devedse Since now output variable support is there in the platform, do you want to refresh your PR accordingly. You can look at this [task]( https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/AzureResourceGroupDeployment/task.json) for an example.

**Comment by devedse** [comment]
It's good to see that it has been included. I'll see if I have some time in the near future to look at the PR. If someone else wants to pick it up in the meanwhile to be a bit faster, I wouldn't mind either :smile:.

**Comment by bansalaseem** [comment]
@chrisrpatterson, @bryanmacfarlane I have assigned the PR to you so that you can share whether output variables is there in the platform or not.

**Comment by JohnGalt1717** [comment]
Where are we at with this? Right now it's impossible to run a docker build and then a docker cp command because we can't get the containerid that was created which is a major blocker.

**Comment by bryanmacfarlane** [comment]
Re-opening.  Output variables are in the platform.  It does have to change per the above comments.

@azooinmyluggage - who should pick up this work in the docker task?  As @JohnGalt1717 pointed out, it would be great to use the container id downstream

**Comment by bansalaseem** [comment]
@vithati can you please pick this up?

**Comment by bansalaseem** [comment]
We are adding support for running N docker commands within a job which will reduce this priority. So I am closing this issue

**Comment by bryanmacfarlane** [comment]
Output variables are still needed for multi phase graph of jobs

See here: https://github.com/Microsoft/vsts-agent/blob/master/docs/preview/yamlgettingstarted-outputvariables.md

**Comment by JohnGalt1717** [comment]
Yes. The example that needs to be handled is being able to execute tests against a pre-built docker image.  To do this you need the container id to execute against for the docker command for the tests.

**Comment by bryanmacfarlane** [comment]
However, note that it shouldn't be implemented like it was in this PR.  We have first class output variable concept in the task json.  We don't need an input which is output variable name.

@JohnGalt1717 or @devedse - I would recommend opening an issue to track doing this properly, link it here and then close this PR.

**Comment by JohnGalt1717** [comment]
@bryanmacfarlane This was the task I opened to track this properly.  This other check in has highjacked it.

**Comment by bansalaseem** [comment]
@JohnGalt1717 Can you please share the link to the issue you created?

**Comment by devedse** [comment]
I'm not fully sure what this item is currently waiting on, is there any knowledge required from my side?

**Comment by bansalaseem** [comment]
@Anumita has picked up adding the output variable support in docker task and she should be able to finish it sometime next week.

**Comment by Anumita** [comment]
I have added output variable support for Docker version 0 task. @devedse , this(https://github.com/Microsoft/vsts-tasks/pull/8152) should solve your issue. Please let us know if you face further issues.

**Comment by bansalaseem** [comment]
Given that Anumita has added the support, I am closing this PR

**Comment by devedse** [comment]
@bansalaseem @Anumita , can my original issue also be solved by this? And is it already released?

https://github.com/Microsoft/vsts-tasks/pull/8183#issuecomment-429102074

**Comment by Anumita** [comment]
@devedse,  it will take about 3 weeks to reach your account as the change is in sprint 142 payload.

**Pull request closed without merging**